hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/IncrSourceHelper.java

```diff
@@ -106,45 +127,108 @@ public static Pair<String, Pair<String, String>> calculateBeginAndEndInstants(Ja
+          beginInstantTime, nthInstant.map(HoodieInstant::getTimestamp).orElse(beginInstantTime),
+          orderColumn, keyColumn, limitColumn);
     } else {
       // when MissingCheckpointStrategy is set to read everything until latest, trigger snapshot query.
       Option<HoodieInstant> lastInstant = activeCommitTimeline.lastInstant();
-      return Pair.of(DataSourceReadOptions.QUERY_TYPE_SNAPSHOT_OPT_VAL(), Pair.of(beginInstantTime, timestampForLastInstant.apply(lastInstant.get())));
+      return new QueryInfo(DataSourceReadOptions.QUERY_TYPE_SNAPSHOT_OPT_VAL(),
+          previousInstantTime, beginInstantTime, lastInstant.get().getTimestamp(),
+          orderColumn, keyColumn, limitColumn);
     }
   }
 
   /**
-   * Validate instant time seen in the incoming row.
+   * Adjust the source dataset to size based batch based on last checkpoint key.
    *
-   * @param row          Input Row
-   * @param instantTime  Hoodie Instant time of the row
-   * @param sinceInstant begin instant of the batch
-   * @param endInstant   end instant of the batch
+   * @param sourceData  Source dataset
+   * @param sourceLimit Max number of bytes to be read from source
+   * @param queryInfo   Query Info
+   * @return end instants along with filtered rows.
    */
-  public static void validateInstantTime(Row row, String instantTime, String sinceInstant, String endInstant) {
-    Objects.requireNonNull(instantTime);
-    ValidationUtils.checkArgument(HoodieTimeline.compareTimestamps(instantTime, HoodieTimeline.GREATER_THAN, sinceInstant),
-        "Instant time(_hoodie_commit_time) in row (" + row + ") was : " + instantTime + "but expected to be between "
-            + sinceInstant + "(excl) - " + endInstant + "(incl)");
-    ValidationUtils.checkArgument(
-        HoodieTimeline.compareTimestamps(instantTime, HoodieTimeline.LESSER_THAN_OR_EQUALS, endInstant),
-        "Instant time(_hoodie_commit_time) in row (" + row + ") was : " + instantTime + "but expected to be between "
-            + sinceInstant + "(excl) - " + endInstant + "(incl)");
+  public static Pair<CloudObjectIncrCheckpoint, Dataset<Row>> filterAndGenerateCheckpointBasedOnSourceLimit(Dataset<Row> sourceData,
+                                                                                                            long sourceLimit, QueryInfo queryInfo,
+                                                                                                            CloudObjectIncrCheckpoint cloudObjectIncrCheckpoint) {
+    if (sourceData.isEmpty()) {
+      LOG.info("Empty source, returning endpoint:" + queryInfo.getEndInstant());
+      return Pair.of(cloudObjectIncrCheckpoint, sourceData);
+    }
+    // Let's persist the dataset to avoid triggering the dag repeatedly
+    sourceData.persist(StorageLevel.MEMORY_AND_DISK());
+    // Set ordering in query to enable batching
+    Dataset<Row> orderedDf = QueryRunner.applyOrdering(sourceData, queryInfo.getOrderByColumns());
```

Comment:
Applying ordering here looks redundant (?) since the entire orderedDf is going to be filtered and after that ordering specified by window ORDER clause will take effect.

Reply:
we need ordering for filtering, but the second ordering I agree might be redundant from the code. I am assuming spark will be able to optimize here.

Reply:
Is ordering preserved after filter? I think let's keep both for now as the impact is low, but also please test if ordering is preserved then we can remove the second window function in a separate patch.

Reply:
AFAIK looking like spark does not guarantee preserving ordering after filtering. so I think its better to keep both.